URLInstrumentation: only set a Task Delegate if there is no Session Delegate

MustafaHaddara opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-swift/issues/744

Majority of the changes are in test/example code. The actual fix is quite small, but relies on calling `task.value(forKey: "session") as? URLSession` to get the associated session on a task. 

## Alternatives
### Intercept Task Creation
One other approach we discussed in the sig was to intercept the task creation and call `objc_setAssociatedObject` on the task, if the session had an active delegate. We already do this here: https://github.com/open-telemetry/opentelemetry-swift/blob/6a2c29d53ff0b543b551b2221538bd3d0206c6d6/Sources/Instrumentation/URLSession/URLSessionInstrumentation.swift#L192-L195

However, that blog of code isn't swizzling the `URLSession.data(for:)` or `URLSession.data(from:)` functions, which are used to create asynchronous tasks. We could start swizzling those functions as well, if we wanted to, but that felt like a riskier change. 

### Swizzle UISession
Honeycomb has [an alternate URL instrumentation implementation](https://github.com/honeycombio/honeycomb-opentelemetry-swift/blob/main/Sources/Honeycomb/Networking/URLSession.swift) that swizzles the `URLSession` and `URLSessionTask` _constructors_ directly. This lets us get a much stronger handle on the sessions and tasks in a wider variety of cases. Switching to that approach would be more costly and time consuming, but may serve us in the long run since the implementation is simpler.

ArielDemarco commented:
The PR looks good to me!

The only thing I’d recommend is adding a few tests to cover the different `NSURLSessionTask` types that can be created (at least the ones this instrumentation handles). I think it’s especially worth testing because the solution relies on KVO of a variable that should exist across the many different `NSURLSessionTask` subclasses, and the test will help ensure nothing breaks across OS versions as time goes by.

That said, the PR is failing due to the examples added. It seems like there's a simple fix needed to wrap some API usages with `@available`

MustafaHaddara commented:
@ArielDemarco thanks for taking a look! I've fixed the tests for tvOS and watchOS, and added some tests for async upload and download tasks with a delegate.
